aptos-move/block-executor/src/executor.rs

```diff
@@ -248,7 +267,13 @@ where
             versioned_data_cache.delete(&k, idx_to_execute);
         }
 
-        last_input_output.record(idx_to_execute, state_view.take_reads(), result);
+        if last_input_output.record(idx_to_execute, state_view.take_reads(), result) {
+            // Optimization for module publishing fallback.
+            // When there is module r/w intersection, can halt parallel execution
+            // and fallback to sequential execution immediately.
+            scheduler.set_done();
```

Comment:
I think scheduler can have a "terminate" method, which internally can set done, and do the loop for condvar resolution.
Setting the done can be done via swap, and whoever actually swaps it from false to true gets to do the resolution (so you won't need the is_first either).